Add Gitea support

mvdkleijn opened the pull request:
## what

Addition of Gitea client in order to close #3538 

**note** this is a draft PR with an initial, naive, probably not complete addition of a Gitea client. This PR is opened in the hopes that other users can provide feedback, help test, etc. since I have a limited amount of time.

## why

See #3538

## tests

none yet

mvdkleijn commented:
Small update for those keeping track: @florianbeisel found my PR and had also started work on his own version for this issue. After contacting me, we decided to work together on this PR.

mvdkleijn commented:
@GenPage @nitrocode @X-Guardian

This PR has been updated with the combined efforts of myself as well as @florianbeisel and we feel it has reached a point where we feel happy it works and is ready to receive feedback or approvals from maintainers and/or users.

Once approved, this will close #3538 by adding nearly full support for Gitea / Forgejo.

**Additional Notes:**
Not implemented: GetTeamNamesForUser()

florianbeisel commented:
I want to extend a huge thank you to @mvdkleijn who I had the pleasure working alongside on this PR. Martijn, It's collaborations like these that make our work not just successful, but also enjoyable.

Looking forward to more great teamwork in the future! Thank you for getting this of the ground!

jamengual commented:
I love the collaboration; thanks, @florianbeisel  and @mvdkleijn, for the work.
Please be patient on the review, this is going to take a while.

jamengual commented:
@mvdkleijn  we are going to need docs/setup guide for gitea, same as we have for the other VCS and include whatever gitea specifics are in it.

bencurio commented on cmd/server.go:
```diff
@@ -156,6 +160,7 @@ const (
 	DefaultExecutableName               = "atlantis"
 	DefaultMarkdownTemplateOverridesDir = "~/.markdown_templates"
 	DefaultGHHostname                   = "github.com"
+	DefaultGiteaBaseURL                 = "https://cloud.gitea.com"
```
I think this URL is wrong, the correct one is: https://gitea.com/ (the index is a redirect to about.gitea.com, but the Gittea instance is on gitea.com)

You might also consider using the URL https://codeberg.org/

mvdkleijn replied:
Corrected. I opted for gitea.com and not something like codeberg since that's more logical IMHO.

mvdkleijn commented:
> @mvdkleijn we are going to need docs/setup guide for gitea, same as we have for the other VCS and include whatever gitea specifics are in it.

On it. Will commit those sometime this weekend.

mvdkleijn commented:
Website docs added to PR

nitrocode commented:
Hi all. The pr says this is a draft and no tests. Is that accurate? If not, could you update this? Tests are always nice when adding a new feature as the feature could break in the future and tests would prevent merging a follow up breaking PR.

florianbeisel commented:
Hi @nitrocode, the description needs to be updated - I think @mvdkleijn will be on this already. From our perspective this has now come out of a pure draft state. The implementation is functional complete and has been tested against a current self hosted Gitea instance and a Forgejo instance. 

While you are right that more tests, more better we opted to concentrate on the tests that are testing the interfaces (`TestPost_UnsupportedGiteaEvent` etc,) first. From our point of view, writing a comprehensive test suite for the client itself would mostly be testing the gitea SDK which we weren't certain would be a real benefit.

mvdkleijn commented:
@nitrocode Sorry, I forgot to update the description. Updated it.

Like @florianbeisel said, I'm looking atthe failed test. I was working on it last night but it was not co-operating. It is clear to me however that it is not a functional issue with our implementation.

I should have some time to work on that today.

Having said all that, the implementation is functionally done and documentation was added / updated as well.

nitrocode commented on go.mod:
```diff
@@ -3,13 +3,14 @@ module github.com/runatlantis/atlantis
 	github.com/cactus/go-statsd-client/v5 v5.1.0
 	github.com/go-ozzo/ozzo-validation v3.6.0+incompatible
-	github.com/go-playground/validator/v10 v10.18.0
+	github.com/go-playground/validator/v10 v10.17.0
```
Why are dependencies getting downgraded?

mvdkleijn replied:
That's a good question. I'll look into that.

mvdkleijn commented:
Thanks @florianbeisel I think I was over-complicating things for myself causing some code blindness in between bouts of my kid not feeling well. :) 

@nitrocode All tests working. All docs present. Dependencies back to what they should be.

mvdkleijn commented:
Just an additional note, the CodeQL check that's failing is doing so based on code we did not touch. It falls under the header "Unchanged files with check annotations (Beta)". So not relevant to this PR in my opinion.

nitrocode commented on cmd/server_test.go:
```diff
@@ -174,6 +179,7 @@ func TestExecute_Defaults(t *testing.T) {
 	strExceptions := map[string]string{
 		GHUserFlag:                       "user",
 		GHTokenFlag:                      "token",
+		GiteaBaseURLFlag:                 "https://gitea-base-url.com",
```
Can we default this to `localhost` instead ? Seems a little risky to put a real domain name when we do not control it

```suggestion
		GiteaBaseURLFlag:                 "http://localhost",
```

nitrocode commented on runatlantis.io/docs/deployment.md:
````diff
@@ -481,6 +514,27 @@ containers:
+#### Gitea
+
+```yaml
+...
````
This will break the yaml color coding if not remove

```suggestion
```

mvdkleijn replied:
Fine by me, but then all the other existing code snippets are also breaking the color coding. :smile:

nitrocode replied:
Oof. Thanks for fixing. At least now we will have a good example to point to help improve docs further. 😁

nitrocode commented on runatlantis.io/docs/server-configuration.md:
````diff
@@ -472,6 +472,43 @@ and set `--autoplan-modules` to `false`.
+  # or
+  ATLANTIS_GITEA_BASE_URL="http://your-gitea.corp:7990/basepath"
+  ```
+  Base URL of Gitea installation. Must include `http://` or `https://`. If using Gitea cloud (gitea.com), do not set.  /Defaults to `https://gitea.com`.
````
```suggestion
  Base URL of Gitea installation. Must include `http://` or `https://`. If using Gitea cloud (gitea.com), do not set. Defaults to `https://gitea.com`.
```

nitrocode commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,468 @@
+	listOptions := gitea.ListPullRequestFilesOptions{
+		ListOptions: gitea.ListOptions{
+			Page:     1,
+			PageSize: 100,
```
Is this the default pagesize?

Do we want this option to be configurable?

According to https://pkg.go.dev/code.gitea.io/sdk/gitea#ListOptions, it seems that it's configurable on the gitea server so some users may need to increase or decrease it.

mvdkleijn replied:
You're right, this should probably be configurable. I'll add a flag to allow for that. The value is dependent on the specific Gitea installation's configuration. I'll apply the default from the Gitea docs.

mvdkleijn replied:
Added --gitea-page-size, defaulting to 30 based on Gitea docs.

nitrocode commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,468 @@
+
+		nextPage = resp.NextPage
+
+		// Emergency break after 500 pages
+		if page >= 500 {
```
Why after 500 ?

mvdkleijn replied:
This is a randomly selected value. It **should** never be hit, but is there to prevent a runaway situation. Not sure if this should be a constant or a configurable value honestly.

nitrocode replied:
Let's start with a constant

mvdkleijn replied:
Change into const with some comments next to it.

nitrocode commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,468 @@
+		opts := gitea.ListIssueCommentOptions{
+			ListOptions: gitea.ListOptions{
+				Page:     nextPage,
+				PageSize: 100,
```
This value seems hard coded. Please move it to a constant or an input flag.

nitrocode commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,468 @@
+	listOptions := gitea.ListPullReviewsOptions{
+		ListOptions: gitea.ListOptions{
+			Page:     1,
+			PageSize: 100,
```
hard coded, please see previous comments

mvdkleijn replied:
Added --gitea-page-size, defaulting to 30 based on Gitea docs.

nitrocode commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,468 @@
+	listOptions := gitea.ListPullReviewsOptions{
+		ListOptions: gitea.ListOptions{
+			Page:     1,
+			PageSize: 100,
```
hard coded, please see previous comments

mvdkleijn replied:
Added --gitea-page-size, defaulting to 30 based on Gitea docs.

nitrocode commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,468 @@
+		nextPage = resp.NextPage
+
+		// Emergency break after 500 pages
+		if page >= 500 {
```
hard coded, please see previous comments

mvdkleijn replied:
Changed into const

nitrocode commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,468 @@
+	opts := gitea.ListLabelsOptions{
+		ListOptions: gitea.ListOptions{
+			Page:     0,
+			PageSize: 100,
```
hard coded, please see previous comments

mvdkleijn replied:
Added --gitea-page-size, defaulting to 30 based on Gitea docs.

nitrocode commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,468 @@
+		nextPage = resp.NextPage
+
+		// Emergency break after 500 pages
+		if page >= 500 {
```
hard coded, please see previous comments

mvdkleijn replied:
Changed into const

nitrocode commented:
How long have you been running atlantis with gitea support in your production instance?

Can you include any screenshots in your PR to ensure this is working as expected?

cc: @runatlantis/core-contributors @runatlantis/maintainers please help review

mvdkleijn commented on runatlantis.io/docs/server-configuration.md:
````diff
@@ -472,6 +472,43 @@ and set `--autoplan-modules` to `false`.
+  # or
+  ATLANTIS_GITEA_BASE_URL="http://your-gitea.corp:7990/basepath"
+  ```
+  Base URL of Gitea installation. Must include `http://` or `https://`. If using Gitea cloud (gitea.com), do not set.  /Defaults to `https://gitea.com`.
````
Redid string to read: `Base URL of Gitea installation. Must include 'http://' or 'https://'. Defaults to 'https://gitea.com' if left empty/absent.`

mvdkleijn commented:
> How long have you been running atlantis with gitea support in your production instance?
> 
> Can you include any screenshots in your PR to ensure this is working as expected?
> 
> cc: @runatlantis/core-contributors @runatlantis/maintainers please help review

How long? Since I started developing the client? A little bit of a strange question for me... :smile:

I don't feel comfortable doing that with my prod stuff, but I could run through some steps with a null resource in a test repo if that's ok?

Myself as well as @florianbeisel have done manual testing against Forgejo and Gitea respectively.

Of course I'd welcome it if someone besides myself and Florian could do a manual test run.

nitrocode commented:
@mvdkleijn you can create an image in your fork https://github.com/mvdkleijn/atlantis by merging your branch `feat-gitea-support` into your fork's head branch `main` which will kick off a github action to build the container and upload it to your fork's github container registry (ghcr). This image can then be used in your environment to beta test your changes.

mvdkleijn commented:
As per @nitrocode suggestion, dev images including Gitea support are now available to test with at https://github.com/mvdkleijn/atlantis/pkgs/container/atlantis

If someone uses it to test please give us your feedback here.

nitrocode commented:
@jamengual @GenPage @lukemassa do we require a new vcs to be deployed in an existing environment, or is there anything else to be addressed, or is this good to merge?

mvdkleijn commented:
> @mvdkleijn is it possible to use this new image in your setup to ensure it works as expected? This may catch some edge cases prior to merge

I'll see if I can use the image locally and possibly provide a screenshot of some basic scenarios with a simple test repo.

florianbeisel commented:
I'll have no way of running any kind of production stuff through it currently but I can set some things up in my homelab I guess.

florianbeisel commented:
![grafik](https://github.com/runatlantis/atlantis/assets/591817/0cc86367-53c7-4c0f-97bb-c4c9b8a97dc2)
![grafik](https://github.com/runatlantis/atlantis/assets/591817/0353f6fe-3494-4132-94de-55c28d4d797a)
![grafik](https://github.com/runatlantis/atlantis/assets/591817/dbe0659d-6252-4a50-b47a-b9c3dcd3ee7e)
![grafik](https://github.com/runatlantis/atlantis/assets/591817/c7cec3eb-aedc-4c0a-b28f-934633b21865)

icyleaf commented on server/events/event_parser.go:
```diff
@@ -610,6 +625,13 @@ func (e *EventParser) ParseGithubRepo(ghRepo *github.Repository) (models.Repo, e
+// returns a repo into the Atlantis model.
+// See EventParsing for return value docs.
+func (e *EventParser) ParseGiteaRepo(repo giteasdk.Repository) (models.Repo, error) {
+	return models.NewRepo(models.Gitea, repo.FullName, repo.CloneURL, e.GithubUser, e.GithubToken)
```
Passed incorrect arguments: 

`GithubUser` -> `GiteaUser`
`GithubToken` -> `GiteaToken`

icyleaf replied:
And variable eventParser in `server/server.go#546` file missing those value above.

mvdkleijn replied:
Corrected. I'll try to get screenshots of a private repo later tonight.

nitrocode commented:
The tests still shows "none yet" in the description.

It would be good to describe what testing has been done for this PR in the PR description.

nitrocode commented:
@florianbeisel in your screenshot, is the above test against gitea?

mvdkleijn commented:
> The tests still shows "none yet" in the description.
> 
> It would be good to describe what testing has been done for this PR in the PR description.

Added test description to PR description

GenPage commented:
We need to hold on this until we merge #4285 from @X-Guardian and then rebase this PR on it before the 0.28 release cycle

nitrocode commented:
@mvdkleijn Thanks for testing and adding that screenshot.

Is that of a private repo? or a public repo? Ideally both types (or if there are additional types in gitea/forejo) are tested.

mvdkleijn commented:
> @mvdkleijn Thanks for testing and adding that screenshot.
> 
> Is that of a private repo? or a public repo? Ideally both types (or if there are additional types in gitea/forejo) are tested.
> 
> 

That's for a public repo, I didn't have time yet to add them for a private one. I can possibly do that later this evening. (depends on the kid 😋)

florianbeisel commented:
> @florianbeisel in your screenshot, is the above test against gitea?

yes this is against a current Gitea installation

mvdkleijn commented:
@nitrocode Here's the private repo. I assume the screenshot is clear like this, otherwise give me a shout.

![Screenshot 2024-02-27 at 16-47-59 test-one](https://github.com/runatlantis/atlantis/assets/431452/43957135-56d8-49d2-b9ec-e82e6127390a)

mvdkleijn commented:
For those keeping track, I had the test docker image rebuilt with the latest state of the PR: ghcr.io/mvdkleijn/atlantis:dev-alpine-7ad19a3

icyleaf commented:
The new container works for my private repo.

nitrocode commented:
Fantastic. Thanks for testing the private repo. Let's fix the merge conflicts and then we can review once more

mvdkleijn commented:
- Resolved merge conflicts
- Updated Gitea client impl to conform to new/updated Client interface (added logger DI)
- Ran local unit tests
- Did a manual test against private repo to ensure everything was still working
- Updated the test image, find it at ghcr.io/mvdkleijn/atlantis:dev-debian or ghcr.io/mvdkleijn/atlantis:dev-alpine

Possibly of interest: the test image has been downloaded 94 time so far...

@nitrocode Thanks for sticking around and guiding this PR in, that helps a lot. :smile:

X-Guardian commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,481 @@
+	token       string
+	pageSize    int
+	ctx         context.Context
+	logger      logging.SimpleLogging
```
You don't need this `logger` property any more.

mvdkleijn replied:
Good point, removed.

X-Guardian commented on server/events/vcs/gitea/client.go:
```diff
@@ -0,0 +1,481 @@
+	}
+	repository, _, err := c.giteaClient.GetRepo(parts[0], parts[1])
+	if err != nil {
+		c.logger.Debug("GET /repos/%v/%v returned an error: %v", parts[0], parts[1], err)
```
```suggestion
		logger.Debug("GET /repos/%v/%v returned an error: %v", parts[0], parts[1], err)
```

X-Guardian reviewed (changes requested):
I would suggest adding some more debug lines to the gitea client code, similar to what was added to the GitHub/Gitlab client code in https://github.com/runatlantis/atlantis/pull/4285. This will help debugging issues in the client if they occur.

X-Guardian reviewed (changes requested):
I would also recommend adding debug logging after every `c.giteaClient` function call, outputting the gitea API endpoint details and the HTTP response code received, the same as the GitHub and GitLab clients do. Example: https://github.com/runatlantis/atlantis/blob/b4ab9c08e24df655c01b221a6cb581c87506b4ed/server/events/vcs/github_client.go#L209-L212

mvdkleijn commented:
> I would also recommend adding debug logging after every `c.giteaClient` function call, outputting the gitea API endpoint details and the HTTP response code received, the same as the GitHub and GitLab clients do. Example:
> 
> https://github.com/runatlantis/atlantis/blob/b4ab9c08e24df655c01b221a6cb581c87506b4ed/server/events/vcs/github_client.go#L209-L212

We're using the Gitea SDK, so we're not making direct HTTP calls. Any errors that are returned by the SDK are either handled or returned.

X-Guardian commented:
Yes, same pattern as the Atlantis GitHub and GitLab clients. Take a look at them.

mvdkleijn commented:
Hmmm... I don't really like the idea of putting the Gitea REST endpoints in debug logging statements which I believe is what you're suggesting. That would mean having to update the statements anytime the API changes right? I feel like that kind of defeats the purpose of using an SDK.

What would be an option... I could check if logger is set to Debug level and in that case use the SDK's SetDebugMode() function? That would put the SDK in debug level?

X-Guardian commented:
The 'SDK' is just a thin wrapper around the Gitea API, exactly the same as the GitLab and GitHub Go packages used by Atlantis. The endpoints will be fixed. When there are issues, without a way of seeing the API return codes, it will be very difficult to work out what is happening. We have been through this pain with GitHub and GitLab. 

The verbose logging within Atlantis commands relies on debug lines being written to the logger irrelevant of whether the server is running with the debug log-level.

icyleaf commented:
Perhaps use the [SetDebugMode](https://pkg.go.dev/code.gitea.io/sdk/gitea#SetDebugMode) method in the Gitea SDK to enable debug mode directly?

mvdkleijn commented:
That's what I suggested earlier but that's apparently not acceptable. I'll revisit this either tonight or tomorrow after I'm done pulling cables.

mvdkleijn commented:
@X-Guardian I still do not agree, but apparently the recommendation is mandatory. I added the requested debug statements.

mvdkleijn commented:
Hey all / @nitrocode is this still waiting for / need something?

nitrocode commented:
Hi @mvdkleijn. Please fix the merge conflict and request a review from x-guardian.

mvdkleijn commented:
@X-Guardian I've added the requested logging, was there anything else that stood out? I'll fix the merge conflict (only small conflict in go.sum) tonight when I have proper access.

florianbeisel commented:
@mvdkleijn: the merge conflict should be fixed now

jamengual reviewed (approved):
LGTM

GenPage commented:
I will finish up the review tonight and coordinate with @X-Guardian

GenPage commented:
> @X-Guardian I still do not agree, but apparently the recommendation is mandatory. I added the requested debug statements.

To clarify, when you use the Atlantis commands in a PR comment, you can pass a "verbose" flag that will append all logging output to the comment response. This includes log events that are debug level, even if the Atlantis server configuration is not set to debug. 

It also appears that `SetDebugMode` doesn't log messages but rather prints to standard output with `fmt.Printf`. This would prevent Atlantis from being able to capture those messages and is why `SetDebugMode` would be an insufficient requirement. @X-Guardian was correct into requiring the additional debug logging in the application implementation.

Otherwise, the code looks good! Thank you both for the time/effort to implement support for Gitea. The community will greatly benefit and appreciate the support.

florianbeisel commented:
Thanks @GenPage and @X-Guardian for the context on the debug statements and where they are being used in practice. This has cleared up a few things for me. I don't want to speak for @mvdkleijn but I still would like to give some final context - though I believe we're now all on the same page here.

I initially had reservations about the approach to debug logging. Our decision to rely on the SetDebugMode function was based on our understanding and discussions at the time. However, I acknowledge that at least my knowledge was incomplete, especially regarding the `verbose` flag.

For me implementing support for Gitea was a significant effort and my biggest contribution to any project really and I am glad to have been a part of it. Y'alls feedback has been invaluable in steering this PR in the right direction (or keeping it on track).

So thank you again for creating a welcoming atmosphere and seeing this through to completion. I will be looking forward to contribute tp the project again should something come up

mvdkleijn commented:
Hi there @GenPage , @X-Guardian , @nitrocode and others,

I'd like to echo @florianbeisel 's sentiment with regards to the friendly reception we received when working on this PR.

One aspect I appreciate for example is that the PR did not "hang in limbo" too long as it were, since the longer a PR takes to get reviewed, the more difficult it is to manage for the creator in my personal experience.

The requirement for and clarification of the logging certainly helps with understanding and as such is definitely appreciated. Sometimes one can get too focused on specific parts too keep an eye on the total.

Though on occasion I had to find time in between family life due to stuff, as no doubt goes for others here too... I feel happy that we could contribute this to the community.

I'd also like to say that working together with @florianbeisel was a very nice experience from my point of view.

We managed to more quickly respond to feedback because of the co-op when one of us was temporarily occupied. The co-op was much appreciated; thanks for the experience Florian.

Like Florian I fully intend to support this code and (time permitting) work on other code as well.

Though maybe I'll restrict myself to before 00:00 at night next time Florian. (^_^)
